core/server/models/plugins/filter.js

```diff
@@ -1,92 +1,175 @@
+    replacement: 'tags.slug',
+    filter: 'posts_tags.sort_order:0+visibility:public'
+}, {
+    key: 'primary_author',
+    replacement: 'users.slug',
+    filter: 'posts_authors.sort_order:0+visibility:public'
+}];
+
+// @TODO: The filter utility lives here temporary.
+const filterUtils = {
     /**
-     * ## Combine Filters
-     * Util to combine the enforced, default and custom filters such that they behave accordingly
-     * @param {String|Object} enforced - filters which must ALWAYS be applied
-     * @param {String|Object} defaults - filters which must be applied if a matching filter isn't provided
-     * @param {...String|Object} [custom] - custom filters which are additional
-     * @returns {*}
+     * ## Get filter keys
+     *
+     * Returns keys used in a query string, e.g.:
+     *
+     * ('featured:true') => ['featured:']
+     * ('page:false+status:published') => ['page:', 'status:']
      */
-    combineFilters: function combineFilters(enforced, defaults, custom /* ...custom */) {
-        custom = Array.prototype.slice.call(arguments, 2);
-
-        // Ensure everything has been run through the gql parser
-        try {
-            enforced = enforced ? (_.isString(enforced) ? gql.parse(enforced) : enforced) : null;
-            defaults = defaults ? (_.isString(defaults) ? gql.parse(defaults) : defaults) : null;
-            custom = _.map(custom, function (arg) {
-                return _.isString(arg) ? gql.parse(arg) : arg;
-            });
-        } catch (err) {
-            throw new common.errors.ValidationError({
-                err: err,
-                property: 'filter',
-                context: common.i18n.t('errors.models.plugins.filter.errorParsing'),
-                help: common.i18n.t('errors.models.plugins.filter.forInformationRead', {url: 'https://api.ghost.org/docs/filter'})
-            });
+    getFilterKeys: (query) => {
+        const tokens = nql(query).lex();
+
+        return tokens
+            .filter(t => t.token === 'PROP')
+            .map(t => t.matched);
+    },
+
+    /**
+     * ## Reduce filters
+     *
+     * Removes filter keys from secondary filter if they are present
+     * in the primary filter, e.g.:
+     *
+     * ('featured:true', 'featured:false') => ''
+     * ('featured:true', 'featured:false,status:published') => 'status:published'
+     */
+    reduceFilters: (primary, secondary) => {
+        if (!primary || !secondary) {
+            return secondary;
         }
 
-        // Merge custom filter options into a single set of statements
-        custom = gql.json.mergeStatements.apply(this, custom);
+        const primaryKeys = filterUtils.getFilterKeys(primary);
+        let reducedFilter = secondary;
 
-        // if there is no enforced or default statements, return just the custom statements;
-        if (!enforced && !defaults) {
-            return custom;
+        primaryKeys.forEach((key) => {
+            if (reducedFilter.match(key)) {
```

Comment:
i think we have to match `key:`, otherwise if you add an alias "tag" -> "tags.slug", it would match "tags".

Reply:
:+1: fa3517b6d7c133d4eb12d7bf845f0f073ab1ad5d

Reply:
👍 Just realised, i added the comment on the wrong match 🤣 they look so similar